In [1]:
# imports
import pandas as pd
import seaborn as sns

In [2]:
# load in Ahana's final merged dataset
total_data = "https://raw.githubusercontent.com/ehuang13/w210-presidential-election/master/data/Totaldata_Jun_11.csv"
merged_df = pd.read_csv(total_data, encoding = "ISO-8859-1")

In [3]:
# checkout merged_df
print("Shape of full dataset: {}".format(merged_df.shape))
merged_df.sample(10)

Shape of full dataset: (18855, 65)


Unnamed: 0         ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY  \
6219         6219  201648477  Total   48477  2016     48     477   
11019       11019  200035033  Total   35033  2000     35      33   
9719         9719  200013315  Total   13315  2000     13     315   
7475         7475  202023029  Total   23029  2020     23      29   
7646         7646  202027097  Total   27097  2020     27      97   
14381       14381  200438085  Total   38085  2004     38      85   
17776       17776  200842051  Total   42051  2008     42      51   
17568       17568  200839067  Total   39067  2008     39      67   
17725       17725  200841021  Total   41021  2008     41      21   
3955         3955  201613225  Total   13225  2016     13     225   

             STNAME            CTYNAME  AA_FEMALE        ...          \
6219          Texas  Washington County        328        ...           
11019    New Mexico        Mora County          4        ...           
9719        Georgia      Wilcox County          8        ...           
7475          Maine  Washington County         81        ...           
7646      Minnesota    Morrison County         86        ...           
14381  North Dakota       Sioux County          2        ...           
17776  Pennsylvania     Fayette County        213        ...           
17568          Ohio    Harrison County         10        ...           
17725        Oregon     Gilliam County          3        ...           
3955        Georgia       Peach County        147        ...           

       NA_MALE%  WA_FEMALE%  WA_MALE%  TOT_FEMALE%  TOT_MALE%  \
6219   0.000233    0.398384  0.388821     0.507325   0.492675   
11019  0.000386    0.484743  0.491696     0.494979   0.505021   
9719   0.000000    0.297159  0.334377     0.448116   0.551884   
7475   0.000095    0.463004  0.450397     0.505653   0.494347   
7646   0.000241    0.480871  0.490458     0.495161   0.504839   
14381  0.000246    0.060889  0.082249     0.467223   0.532777   
17776  0.000121    0.483808  0.456103     0.511469   0.488531   
17568  0.000126    0.488649  0.473304     0.506572   0.493428   
17725  0.002753    0.479075  0.486784     0.495044   0.504956   
3955   0.000333    0.258474  0.263284     0.516985   0.483015   

       TOT_POP_LESS19%  TOT_POP_20to39%  TOT_POP_40to59%  TOT_POP_Above60%  \
6219          0.259505         0.222183         0.252383          0.265928   
11019         0.299923         0.208382         0.284473          0.207223   
9719          0.254609         0.295536         0.275942          0.173913   
7475          0.210702         0.198666         0.258908          0.331724   
7646          0.254605         0.220989         0.252495          0.271911   
14381         0.414682         0.283820         0.209428          0.092070   
17776         0.231294         0.230709         0.300723          0.237273   
17568         0.248789         0.206654         0.307025          0.237532   
17725         0.210903         0.170154         0.326542          0.292401   
3955          0.266504         0.276791         0.258363          0.198342   

       Unemployment_Rate  
6219                 5.1  
11019               10.5  
9719                 4.8  
7475                 NaN  
7646                 NaN  
14381                6.7  
17776                6.5  
17568                7.2  
17725                4.3  
3955                 7.0  

[10 rows x 65 columns]

In [4]:
# see all columns in merged dataframe
print(merged_df.columns)

Index(['Unnamed: 0', 'ID', 'AGEGRP', 'FIPS_x', 'YEAR', 'STATE', 'COUNTY',
       'STNAME', 'CTYNAME', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE',
       'H_FEMALE', 'H_MALE', 'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE',
       'SUMLEV', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE',
       'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19',
       'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39',
       'TOT_POP_40to59', 'TOT_MALE_40to59', 'TOT_FEMALE_40to59',
       'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'fips',
       'republican_candidate', 'democrat_candidate', 'republican_votes',
       'democrat_votes', 'county_totalvotes', 'winning candidate',
       'winning party', 'avg_wage_salary', 'AA_FEMALE%', 'AA_MALE%',
       'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%',
       'IA_MALE%', 'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%',
       'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 'TOT_POP_20to39%',


In [5]:
# see how many nulls there are within dataframe
merged_df.isnull().sum()

Unnamed: 0                 0
ID                         0
AGEGRP                     0
FIPS_x                     0
YEAR                       0
STATE                      0
COUNTY                     0
STNAME                     0
CTYNAME                    0
AA_FEMALE                  0
AA_MALE                    0
BA_FEMALE                  0
BA_MALE                    0
H_FEMALE                   0
H_MALE                     0
IA_FEMALE                  0
IA_MALE                    0
NA_FEMALE                132
NA_MALE                    0
SUMLEV                     0
TOT_FEMALE                 0
TOT_MALE                   0
TOT_POP                    0
WA_FEMALE                  0
WA_MALE                    0
TOT_POP_LESS19             0
TOT_MALE_LESS19            0
TOT_FEMALE_LESS19          0
TOT_POP_20to39             0
TOT_MALE_20to39            0
                        ... 
TOT_MALE_Above60           0
TOT_FEMALE_Above60         0
fips                    3280
republican_can

### One hot encode "winning party" column to binary feature
* republican = 1
* democrat = 0

In [6]:
winning_party_binary = pd.get_dummies(merged_df["winning party"], drop_first = True)
merged_df["winning_party_binary"] = winning_party_binary

In [7]:
merged_df.sample(10)

Unnamed: 0         ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY  \
17855       17855  200845065  Total   45065  2008     45      65   
15907       15907  200813117  Total   13117  2008     13     117   
7870         7870  202029209  Total   29209  2020     29     209   
6172         6172  201648383  Total   48383  2016     48     383   
8067         8067  202034019  Total   34019  2020     34      19   
16914       16914  200828043  Total   28043  2008     28      43   
11078       11078  200036089  Total   36089  2000     36      89   
687           687   20201111  Total    1111  2020      1     111   
7175         7175  202020007  Total   20007  2020     20       7   
5382         5382  201637069  Total   37069  2016     37      69   

               STNAME              CTYNAME  AA_FEMALE          ...           \
17855  South Carolina     McCormick County         20          ...            
15907         Georgia       Forsyth County       3973          ...            
7870         Missouri         Stone County         83          ...            
6172            Texas        Reagan County         14          ...            
8067       New Jersey     Hunterdon County       2711          ...            
16914     Mississippi       Grenada County         38          ...            
11078        New York  St. Lawrence County        396          ...            
687           Alabama      Randolph County         53          ...            
7175           Kansas        Barber County         15          ...            
5382   North Carolina      Franklin County        230          ...            

       WA_FEMALE%  WA_MALE%  TOT_FEMALE%  TOT_MALE%  TOT_POP_LESS19%  \
17855    0.218855  0.257922     0.454241   0.545759         0.169455   
15907    0.453519  0.455358     0.500442   0.499558         0.321640   
7870     0.493590  0.472550     0.510126   0.489874         0.181549   
6172     0.442714  0.484688     0.479937   0.520063         0.332365   
8067     0.460389  0.450046     0.504731   0.495269         0.217474   
16914    0.299978  0.275845     0.530737   0.469263         0.286053   
11078    0.477011  0.474902     0.492003   0.507997         0.282831   
687      0.396040  0.382354     0.513487   0.486513         0.241760   
7175     0.462880  0.488596     0.485465   0.514535         0.245751   
5382     0.346775  0.350125     0.503820   0.496180         0.257908   

       TOT_POP_20to39%  TOT_POP_40to59%  TOT_POP_Above60%  Unemployment_Rate  \
17855         0.227809         0.295906          0.306829               10.8   
15907         0.259023         0.295273          0.124064                4.8   
7870          0.163123         0.249047          0.406281                NaN   
6172          0.263464         0.254224          0.149947                6.2   
8067          0.206384         0.308402          0.267741                NaN   
16914         0.236400         0.281845          0.195702                7.3   
11078         0.287370         0.258216          0.171584                5.7   
687           0.218790         0.261870          0.277580                NaN   
7175          0.210868         0.225850          0.317531                NaN   
5382          0.229993         0.297197          0.214902                5.2   

       winning_party_binary  
17855                     0  
15907                     1  
7870                      0  
6172                      1  
8067                      0  
16914                     1  
11078                     0  
687                       0  
7175                      0  
5382                      1  

[10 rows x 66 columns]

In [8]:
merged_df[["winning party", "winning_party_binary"]].sample(10)

winning party  winning_party_binary
1078       democrat                     0
14098    republican                     1
11176    republican                     1
822             NaN                     0
3978     republican                     1
5281       democrat                     0
11077      democrat                     0
13765    republican                     1
1061     republican                     1
7689            NaN                     0

### Explore if we could drop null values
* merged_df before dropping null values has: 18,855 rows and 66 columns
* merged_df after dropping null values has: 15,437 rows and 66 columns

In [344]:
merged_df2 = merged_df.dropna()
merged_df2.shape

(15437, 66)

In [345]:
merged_df2.sample(10)

Unnamed: 0         ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY  \
15276       15276  200451079  Total   51079  2004     51      79   
14300       14300  200437123  Total   37123  2004     37     123   
4474         4474  201621047  Total   21047  2016     21      47   
18359       18359   20085065  Total    5065  2008      5      65   
1469         1469  201219097  Total   19097  2012     19      97   
13308       13308  200420197  Total   20197  2004     20     197   
3446         3446  201250017  Total   50017  2012     50      17   
17484       17484  200838005  Total   38005  2008     38       5   
16914       16914  200828043  Total   28043  2008     28      43   
4380         4380  201620069  Total   20069  2016     20      69   

               STNAME            CTYNAME  AA_FEMALE          ...           \
15276        Virginia      Greene County         79          ...            
14300  North Carolina  Montgomery County        240          ...            
4474         Kentucky   Christian County        554          ...            
18359        Arkansas       Izard County         23          ...            
1469             Iowa     Jackson County         28          ...            
13308          Kansas   Wabaunsee County          8          ...            
3446          Vermont      Orange County         86          ...            
17484    North Dakota      Benson County          3          ...            
16914     Mississippi     Grenada County         38          ...            
4380           Kansas        Gray County         14          ...            

       WA_FEMALE%  WA_MALE%  TOT_FEMALE%  TOT_MALE%  TOT_POP_LESS19%  \
15276    0.458656  0.454518     0.503868   0.496132         0.288661   
14300    0.381804  0.381694     0.499633   0.500367         0.270587   
4474     0.339430  0.388626     0.473146   0.526854         0.304787   
18359    0.473214  0.492031     0.485316   0.514684         0.214950   
1469     0.493494  0.484567     0.504489   0.495511         0.255497   
13308    0.483799  0.494304     0.493564   0.506436         0.282882   
3446     0.486959  0.486613     0.501226   0.498774         0.236225   
17484    0.218165  0.240461     0.494389   0.505611         0.373934   
16914    0.299978  0.275845     0.530737   0.469263         0.286053   
4380     0.486718  0.487213     0.499258   0.500742         0.326019   

       TOT_POP_20to39%  TOT_POP_40to59%  TOT_POP_Above60%  Unemployment_Rate  \
15276         0.270792         0.290040          0.150507                3.2   
14300         0.260964         0.275509          0.192941                7.1   
4474          0.353194         0.188460          0.153559                6.2   
18359         0.204250         0.284386          0.296414                6.7   
1469          0.194119         0.302048          0.248336                5.5   
13308         0.192484         0.312176          0.212457                4.9   
3446          0.213355         0.326286          0.224134                4.7   
17484         0.205297         0.241808          0.178962                6.4   
16914         0.236400         0.281845          0.195702                7.3   
4380          0.228345         0.248144          0.197492                2.3   

       winning_party_binary  
15276                     1  
14300                     1  
4474                      1  
18359                     1  
1469                      0  
13308                     1  
3446                      0  
17484                     0  
16914                     1  
4380                      1  

[10 rows x 66 columns]

### Load and merge senate and house election datasets

In [350]:
# load senate and house datasets
senate = "https://raw.githubusercontent.com/ehuang13/w210-presidential-election/master/data/senate.csv"
house = "https://raw.githubusercontent.com/ehuang13/w210-presidential-election/master/data/house.csv"

senate_df = pd.read_csv(senate, encoding = "ISO-8859-1")
house_df = pd.read_csv(house, encoding = "ISO-8859-1")

### Clean up senate election data

In [351]:
# check out senate election dataset
print("Senate data dimensions: {}".format(senate_df.shape))
senate_df.sample(5)

Senate data dimensions: (3421, 18)


year       state state_po  state_fips  state_cen  state_ic     office  \
8     1976  California       CA           6         93        71  US Senate   
672   1986    Colorado       CO           8         84        62  US Senate   
3190  2016      Nevada       NV          32         88        65  US Senate   
202   1978   Minnesota       MN          27         41        33  US Senate   
2163  2006    Maryland       MD          24         52        52  US Senate   

       district stage  special         candidate        party  writein   mode  \
8     statewide   gen    False        Omari Musa  independent    False  total   
672   statewide   gen    False  Timothy E. Wirth     democrat    False  total   
3190  statewide   gen    False       Tony Gumina         none    False  total   
202   statewide   gen    False     Jean T. Brust      workers    False  total   
2163  statewide   gen    False               NaN          NaN     True  total   

      candidatevotes  totalvotes  unofficial     version  
8              31629     7470586       False  20171011.0  
672           529449     1060765       False  20171011.0  
3190           10740     1108294       False  20171011.0  
202             3891     1580622       False  20171011.0  
2163             120     1781139       False  20171011.0

In [352]:
# keep only election years data for senate and house elections
election_years = [2000, 2004, 2008, 2012, 2016]

senate_df = senate_df[senate_df["year"].isin(election_years)]
print("Senate data dimensions: {}".format(senate_df.shape))

Senate data dimensions: (904, 18)


In [113]:
senate_df.head(10)

year       state state_po  state_fips  state_cen  state_ic     office  \
1631  2000     Arizona       AZ           4         86        61  US Senate   
1632  2000     Arizona       AZ           4         86        61  US Senate   
1633  2000     Arizona       AZ           4         86        61  US Senate   
1634  2000     Arizona       AZ           4         86        61  US Senate   
1635  2000  California       CA           6         93        71  US Senate   
1636  2000  California       CA           6         93        71  US Senate   
1637  2000  California       CA           6         93        71  US Senate   
1638  2000  California       CA           6         93        71  US Senate   
1639  2000  California       CA           6         93        71  US Senate   
1640  2000  California       CA           6         93        71  US Senate   

       district stage  special                 candidate  \
1631  statewide   gen    False              William Toel   
1632  statewide   gen    False              Vance Hansen   
1633  statewide   gen    False                   Jon Kyl   
1634  statewide   gen    False          Barry J. Hess II   
1635  statewide   gen    False  Jose Luis "Joe" Camahort   
1636  statewide   gen    False      Medea Susan Benjamin   
1637  statewide   gen    False              Tom Campbell   
1638  statewide   gen    False          Dianne Feinstein   
1639  statewide   gen    False             Brian M. Rees   
1640  statewide   gen    False       Diane Beall Templin   

                     party  writein   mode  candidatevotes  totalvotes  \
1631           independent    False  total          109230     1397076   
1632                 green    False  total          108926     1397076   
1633            republican    False  total         1108196     1397076   
1634           libertarian    False  total           70724     1397076   
1635                reform    False  total           96552    10623608   
1636                 green    False  total          326828    10623608   
1637            republican    False  total         3886853    10623608   
1638              democrat    False  total         5932522    10623608   
1639           natural law    False  total           58537    10623608   
1640  american independent    False  total          134598    10623608   

      unofficial     version  
1631       False  20171011.0  
1632       False  20171011.0  
1633       False  20171011.0  
1634       False  20171011.0  
1635       False  20171011.0  
1636       False  20171011.0  
1637       False  20171011.0  
1638       False  20171011.0  
1639       False  20171011.0  
1640       False  20171011.0

In [353]:
# for each senate election in the state, find the winning party

# group each election by year, state, and party
senate_modified = senate_df.groupby(["year", "state", "state_fips", "party"])["party", "candidatevotes"].max().reset_index()

# sort candidate votes by descending to find the candidates with the most votes
senate_modified = senate_modified.sort_values(by=["year", "state", "state_fips", "candidatevotes"], ascending=False)

# get the party with the most votes by election year and state
senate_modified = senate_modified.groupby(["year", "state", "state_fips"]).first().reset_index()

In [354]:
# check out modified senate dataset
senate_modified.head(5)

year        state  state_fips       party  candidatevotes
0  2000      Arizona           4  republican         1108196
1  2000   California           6    democrat         5932522
2  2000  Connecticut           9    democrat          828902
3  2000     Delaware          10    democrat          181566
4  2000      Florida          12    democrat         2989487

In [355]:
# need to recode winning party to republican or democrat
senate_modified["party"].unique()

array(['republican', 'democrat', 'democrat (not identified on ballot)',
       'democratic-farmer-labor', 'democratic-nonpartisan league',
       'independent for Maine', 'independent'], dtype=object)

In [356]:
# code other democratic affiliated parties to democrat
senate_modified = senate_modified.replace(["democrat (not identified on ballot)", "democratic-farmer-labor", 
                                            "democratic-nonpartisan league"], "democrat")

# code indepdent affiliated parties to independent
senate_modified = senate_modified.replace("independent for Maine", "independent")

# filter out independent party elections
senate_modified = senate_modified[senate_modified.party != "independent"]

In [357]:
# verify only republican and democrat winning elections remain
senate_modified["party"].unique()

array(['republican', 'democrat'], dtype=object)

In [358]:
# create one hot encoded column for winning party
senate_binary = pd.get_dummies(senate_modified["party"], drop_first = True)
senate_modified["senate_winning_party"] = senate_binary

In [406]:
print("Modified senate dataset dimensions: {}".format(senate_modified.shape))
senate_modified.head(5)

Modified senate dataset dimensions: (166, 6)


year        state  state_fips       party  candidatevotes  \
0  2000      Arizona           4  republican         1108196   
1  2000   California           6    democrat         5932522   
2  2000  Connecticut           9    democrat          828902   
3  2000     Delaware          10    democrat          181566   
4  2000      Florida          12    democrat         2989487   

   senate_winning_party  
0                     1  
1                     0  
2                     0  
3                     0  
4                     0

### Clean up house election data

In [360]:
# checkout house election dataset
print("House data dimensions: {}".format(house_df.shape))
house_df.head(10)

House data dimensions: (29636, 19)


year    state state_po  state_fips  state_cen  state_ic    office  \
0  1976  Alabama       AL           1         63        41  US House   
1  1976  Alabama       AL           1         63        41  US House   
2  1976  Alabama       AL           1         63        41  US House   
3  1976  Alabama       AL           1         63        41  US House   
4  1976  Alabama       AL           1         63        41  US House   
5  1976  Alabama       AL           1         63        41  US House   
6  1976  Alabama       AL           1         63        41  US House   
7  1976  Alabama       AL           1         63        41  US House   
8  1976  Alabama       AL           1         63        41  US House   
9  1976  Alabama       AL           1         63        41  US House   

   district stage runoff  special                      candidate        party  \
0         1   gen  False    False                 Bill Davenport     democrat   
1         1   gen  False    False                   Jack Edwards   republican   
2         1   gen  False    False                            NaN          NaN   
3         2   gen  False    False               J. Carole Keahey     democrat   
4         2   gen  False    False                            NaN          NaN   
5         2   gen  False    False  William L. ""Bill"" Dickinson   republican   
6         3   gen  False    False                   Bill Nichols     democrat   
7         3   gen  False    False                            NaN          NaN   
8         3   gen  False    False                 Ogburn Gardner  prohibition   
9         4   gen  False    False                     Tom Bevill     democrat   

   writein   mode  candidatevotes  totalvotes  unofficial   version  
0    False  total           58906      157170       False  20200424  
1    False  total           98257      157170       False  20200424  
2     True  total               7      157170       False  20200424  
3    False  total           66288      156362       False  20200424  
4     True  total               5      156362       False  20200424  
5    False  total           90069      156362       False  20200424  
6    False  total          106935      108048       False  20200424  
7     True  total               2      108048       False  20200424  
8    False  total            1111      108048       False  20200424  
9    False  total          141490      176022       False  20200424

In [109]:
# filter for only election years data for senate and house elections
house_df = house_df[house_df["year"].isin(election_years)]
print("House data dimensions: {}".format(house_df.shape))

House data dimensions: (7196, 19)


In [361]:
# house election dataset seems to be structured similarly to senate election dataset
house_modified = house_df.groupby(["year", "state", "state_fips", "party"])["party", "candidatevotes"].max().reset_index()

# sort candidate votes by descending to find the candidates with the most votes
house_modified = house_modified.sort_values(by=["year", "state", "state_fips", "candidatevotes"], ascending=False)

# get the party with the most votes by election year and state
house_modified = house_modified.groupby(["year", "state", "state_fips"]).first().reset_index()

In [362]:
house_modified.head(5)

year       state  state_fips       party  candidatevotes
0  1976     Alabama           1    democrat          141490
1  1976      Alaska           2  republican           83722
2  1976     Arizona           4    democrat          106054
3  1976    Arkansas           5    democrat          144780
4  1976  California           6  republican          173576

In [363]:
# need to recode winning party to republican or democrat
house_modified["party"].unique()

array(['democrat', 'republican', 'independent', 'democratic-farmer-labor'],
      dtype=object)

In [364]:
# code other democratic affiliated parties to democrat
house_modified = house_modified.replace("democratic-farmer-labor", "democrat")

# filter out independent party elections
house_modified = house_modified[house_modified.party != "independent"]

In [365]:
# verify only republican and democrat winning elections remain
house_modified["party"].unique()

array(['democrat', 'republican'], dtype=object)

In [366]:
# create one hot encoded column for winning party
house_binary = pd.get_dummies(house_modified["party"], drop_first = True)
house_modified["house_winning_party"] = house_binary

In [367]:
print("Modified house dataset dimensions: {}".format(house_modified.shape))
house_modified.head(10)

Modified house dataset dimensions: (1092, 6)


year        state  state_fips       party  candidatevotes  \
0  1976      Alabama           1    democrat          141490   
1  1976       Alaska           2  republican           83722   
2  1976      Arizona           4    democrat          106054   
3  1976     Arkansas           5    democrat          144780   
4  1976   California           6  republican          173576   
5  1976     Colorado           8  republican          126784   
6  1976  Connecticut           9  republican          157009   
7  1976     Delaware          10  republican          110636   
8  1976      Florida          12    democrat          199031   
9  1976      Georgia          13    democrat          113245   

   house_winning_party  
0                    0  
1                    1  
2                    0  
3                    0  
4                    1  
5                    1  
6                    1  
7                    1  
8                    0  
9                    0

### Merge senate and house election dataset 

In [395]:
# sort candidate votes by descending to find the candidates with the most votes by district
house_senate_merged = pd.merge(house_modified[["year", "state", "state_fips", "house_winning_party"]],
                               senate_modified[["year", "state", "state_fips", "senate_winning_party"]],
                               how="left", left_on=["year", "state_fips"], right_on=["year", "state_fips"])

In [396]:
house_senate_merged.head()

year     state_x  state_fips  house_winning_party state_y  \
0  1976     Alabama           1                    0     NaN   
1  1976      Alaska           2                    1     NaN   
2  1976     Arizona           4                    0     NaN   
3  1976    Arkansas           5                    0     NaN   
4  1976  California           6                    1     NaN   

   senate_winning_party  
0                   NaN  
1                   NaN  
2                   NaN  
3                   NaN  
4                   NaN

In [397]:
# verify data types
house_senate_merged.dtypes

year                      int64
state_x                  object
state_fips                int64
house_winning_party       uint8
state_y                  object
senate_winning_party    float64
dtype: object

### Merge senate and house election dataset to Ahana's total dataset

In [398]:
# verify dataset dimensions before merge
print("Ahana's dataset dimensions: {}".format(merged_df2.shape))

Ahana's dataset dimensions: (15437, 66)


In [414]:
# merge dataframes on election year and state fips code
combined_df = pd.merge(merged_df2, house_senate_merged, how="left", 
                       left_on=["YEAR", "STATE"], right_on=["year", "state_fips"])

In [421]:
print("Combined dataset dimensions: {}".format(combined_df.shape))
combined_df.head()

Combined dataset dimensions: (15437, 72)


Unnamed: 0        ID AGEGRP  FIPS_x  YEAR  STATE  COUNTY   STNAME  \
0           0  20121001  Total    1001  2012      1       1  Alabama   
1           1  20121003  Total    1003  2012      1       3  Alabama   
2           2  20121005  Total    1005  2012      1       5  Alabama   
3           3  20121007  Total    1007  2012      1       7  Alabama   
4           4  20121009  Total    1009  2012      1       9  Alabama   

          CTYNAME  AA_FEMALE          ...           TOT_POP_40to59%  \
0  Autauga County        296          ...                  0.282427   
1  Baldwin County        833          ...                  0.282293   
2  Barbour County         59          ...                  0.278741   
3     Bibb County         10          ...                  0.280955   
4   Blount County         82          ...                  0.276594   

   TOT_POP_Above60%  Unemployment_Rate  winning_party_binary    year  state_x  \
0          0.171659                6.9                     1  2012.0  Alabama   
1          0.234071                7.5                     1  2012.0  Alabama   
2          0.208233               11.5                     0  2012.0  Alabama   
3          0.187128                8.5                     1  2012.0  Alabama   
4          0.210691                6.9                     1  2012.0  Alabama   

   state_fips  house_winning_party  state_y  senate_winning_party  
0         1.0                  0.0      NaN                   NaN  
1         1.0                  0.0      NaN                   NaN  
2         1.0                  0.0      NaN                   NaN  
3         1.0                  0.0      NaN                   NaN  
4         1.0                  0.0      NaN                   NaN  

[5 rows x 72 columns]

In [417]:
combined_df.columns

Index(['Unnamed: 0', 'ID', 'AGEGRP', 'FIPS_x', 'YEAR', 'STATE', 'COUNTY',
       'STNAME', 'CTYNAME', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE',
       'H_FEMALE', 'H_MALE', 'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE',
       'SUMLEV', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE',
       'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19',
       'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39',
       'TOT_POP_40to59', 'TOT_MALE_40to59', 'TOT_FEMALE_40to59',
       'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'fips',
       'republican_candidate', 'democrat_candidate', 'republican_votes',
       'democrat_votes', 'county_totalvotes', 'winning candidate',
       'winning party', 'avg_wage_salary', 'AA_FEMALE%', 'AA_MALE%',
       'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%',
       'IA_MALE%', 'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%',
       'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 'TOT_POP_20to39%',


In [422]:
# drop unnecessary columns
dropCombinedCols = ["SUMLEV", "AGEGRP", "FIPS_x", "year", "state_x", "state_fips", "state_y"]
combined_df = combined_df.drop(columns = dropCombinedCols)

In [423]:
combined_df.columns

Index(['Unnamed: 0', 'ID', 'YEAR', 'STATE', 'COUNTY', 'STNAME', 'CTYNAME',
       'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE', 'H_FEMALE', 'H_MALE',
       'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE', 'TOT_FEMALE',
       'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE', 'TOT_POP_LESS19',
       'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19', 'TOT_POP_20to39',
       'TOT_MALE_20to39', 'TOT_FEMALE_20to39', 'TOT_POP_40to59',
       'TOT_MALE_40to59', 'TOT_FEMALE_40to59', 'TOT_POP_Above60',
       'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'fips',
       'republican_candidate', 'democrat_candidate', 'republican_votes',
       'democrat_votes', 'county_totalvotes', 'winning candidate',
       'winning party', 'avg_wage_salary', 'AA_FEMALE%', 'AA_MALE%',
       'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%',
       'IA_MALE%', 'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%',
       'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 'TOT_POP_20to39%',
       'TOT_POP_40to59%', 'TOT

In [427]:
# rename columns
renameCombinedCols = ['ID', 'YEAR_FIPS', 'YEAR', 'STATE_FIPS', 'COUNTY_FIPS', 'STATE', 'COUNTY',
       'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE', 'H_FEMALE', 'H_MALE',
       'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE', 'TOT_FEMALE',
       'TOT_MALE', 'TOT_POP', 'WA_FEMALE', 'WA_MALE', 'TOT_POP_LESS19',
       'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19', 'TOT_POP_20to39',
       'TOT_MALE_20to39', 'TOT_FEMALE_20to39', 'TOT_POP_40to59',
       'TOT_MALE_40to59', 'TOT_FEMALE_40to59', 'TOT_POP_Above60',
       'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'FIPS',
       'REP_CANDIDATE', 'DEM_CANDIDATE', 'REP_VOTES',
       'DEM_VOTES', 'COUNTY_TOTALVOTES', 'WINNING_CANDIDATE',
       'WINNING_PARTY', 'AVG_WAGE_SALARY', 'AA_FEMALE%', 'AA_MALE%',
       'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%',
       'IA_MALE%', 'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%',
       'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 'TOT_POP_20to39%',
       'TOT_POP_40to59%', 'TOT_POP_Above60%', 'UNEMPLOYMENT_RATE',
       'WINNING_PARTY_BINARY', 'HOUSE_WINNING_BINARY', 'SENATE_WINNING_BINARY']

combined_df.columns = renameCombinedCols

# reorder columns
combined_df = combined_df[['ID', 'YEAR_FIPS', 'YEAR', 'STATE_FIPS', 'COUNTY_FIPS', 'STATE', 'COUNTY', 'FIPS',
                            'REP_CANDIDATE', 'DEM_CANDIDATE', 'REP_VOTES', 'DEM_VOTES', 'COUNTY_TOTALVOTES', 'WINNING_CANDIDATE',
                            'WINNING_PARTY', 'WINNING_PARTY_BINARY', 'HOUSE_WINNING_BINARY', 'SENATE_WINNING_BINARY','UNEMPLOYMENT_RATE',
                            'AVG_WAGE_SALARY', 'AA_FEMALE', 'AA_MALE', 'BA_FEMALE', 'BA_MALE', 'H_FEMALE', 'H_MALE',
                            'IA_FEMALE', 'IA_MALE', 'NA_FEMALE', 'NA_MALE', 'TOT_FEMALE', 'TOT_MALE', 'TOT_POP', 
                            'WA_FEMALE', 'WA_MALE', 'TOT_POP_LESS19', 'TOT_MALE_LESS19', 'TOT_FEMALE_LESS19', 
                            'TOT_POP_20to39', 'TOT_MALE_20to39', 'TOT_FEMALE_20to39', 'TOT_POP_40to59', 'TOT_MALE_40to59', 
                            'TOT_FEMALE_40to59', 'TOT_POP_Above60', 'TOT_MALE_Above60', 'TOT_FEMALE_Above60', 'AA_FEMALE%', 
                            'AA_MALE%', 'BA_FEMALE%', 'BA_MALE%', 'H_FEMALE%', 'H_MALE%', 'IA_FEMALE%', 'IA_MALE%', 
                            'NA_FEMALE%', 'NA_MALE%', 'WA_FEMALE%', 'WA_MALE%', 'TOT_FEMALE%', 'TOT_MALE%', 'TOT_POP_LESS19%', 
                            'TOT_POP_20to39%', 'TOT_POP_40to59%', 'TOT_POP_Above60%']]

In [429]:
# check out final combined dataset
combined_df.head()

ID  YEAR_FIPS  YEAR  STATE_FIPS  COUNTY_FIPS    STATE          COUNTY  \
0   0   20121001  2012           1            1  Alabama  Autauga County   
1   1   20121003  2012           1            3  Alabama  Baldwin County   
2   2   20121005  2012           1            5  Alabama  Barbour County   
3   3   20121007  2012           1            7  Alabama     Bibb County   
4   4   20121009  2012           1            9  Alabama   Blount County   

     FIPS REP_CANDIDATE DEM_CANDIDATE        ...         NA_FEMALE%  NA_MALE%  \
0  1001.0   Mitt Romney  Barack Obama        ...           0.000310  0.000511   
1  1003.0   Mitt Romney  Barack Obama        ...           0.000300  0.000399   
2  1005.0   Mitt Romney  Barack Obama        ...           0.000622  0.001061   
3  1007.0   Mitt Romney  Barack Obama        ...           0.000219  0.001137   
4  1009.0   Mitt Romney  Barack Obama        ...           0.000436  0.000627   

   WA_FEMALE%  WA_MALE% TOT_FEMALE%  TOT_MALE%  TOT_POP_LESS19%  \
0    0.402747  0.390054    0.512912   0.487088         0.293056   
1    0.446461  0.429892    0.510581   0.489419         0.253453   
2    0.237175  0.271277    0.469301   0.530699         0.242225   
3    0.371240  0.394981    0.463799   0.536201         0.248820   
4    0.487947  0.477838    0.505255   0.494745         0.271033   

   TOT_POP_20to39%  TOT_POP_40to59%  TOT_POP_Above60%  
0         0.252858         0.282427          0.171659  
1         0.230183         0.282293          0.234071  
2         0.270801         0.278741          0.208233  
3         0.283097         0.280955          0.187128  
4         0.241682         0.276594          0.210691  

[5 rows x 65 columns]

### Save one hot encoded dataframe to csv file

In [430]:
# output one hot encoded dataframe with null values to csv file
combined_df.to_csv("combined_jun13.csv")